### ISSUE TITLE
Express Checkout: amount displayed does not match charged amount

### ISSUE THREAD

**Issue by jonas-hoebenreich** [woocommerce/woocommerce-gateway-stripe#2304]
**Describe the bug**
I am currently trying to make a plugin of mine (https://wordpress.org/plugins/wc-donation-platform/) compatible with Stripe Express Checkout. In the process, I've stumbled across an issue that is causing an incorrect amount to be charged.

When Express Checkout is enabled on the product page, every time the product selection field is changed, the total amount is calculated using an AJAX request (stripe-payment-request.js, wc-ajax=wc_stripe_get_selected_product_data).
When someone clicks on the payment button, the total amount is calculated one more time (stripe-payment-request.js, wc-ajax=wc_stripe_add_to_cart).
The payment modal displays the amount calculated using wc_stripe_get_selected_product_data, but the amount calculated using wc_stripe_add_to_cart is actually charged.
Normally this is not a problem, because the two amounts are the same, but if the amounts differ the user's trust in the Google Pay modal is betrayed.

**To Reproduce**
1. Activate Stripe Express Checkout for Product Pages
2. Add the following PHP code:
```
add_action( 'woocommerce_before_calculate_totals', function($cart_object) {
	foreach ( $cart_object->cart_contents as $key => $value ) {
		$value['data']->set_price(9999);
	}
} );
```

3. Click the Google Pay button on a product page. In this example, the product costs $1 (Screenshot 1)
4. The wrong amount is charged (Screenshot 2)

**Expected behavior**
The amount displayed in the Google Pay window matches the charged amount. At the very least, some form of error message should pop up.

**Screenshots**
Screenshot 1:
![Clipboard01](https://user-images.githubusercontent.com/64426524/151673882-ba0ef0e8-8137-43f8-b852-b604d5d79382.jpg)

Screenshot 2:
![Clipboard02](https://user-images.githubusercontent.com/64426524/151673901-2d4483ea-e3dc-4ad0-826c-897f89fd9a1a.jpg)

**Environment (please complete the following information):**
 - WordPress Version: 5.9
 - WooCommerce Version: 6.1.1
 - Stripe Plugin Version: 6.1.0
 - Browser [e.g. chrome, safari] and Version: Chrome 97.0.4692.71

**Additional context**
By the way, the Express Checkout handling is implemented much more elegantly by the official PayPal plugin (the amount is calculated only once when clicking the button, which saves server resources.)
In addition, it is easier to pass additional individual parameters, because PayPal sends all parameters matching "[name^='attribute_']". Stripe uses $(".variations_form").find(".variations select") here, which makes it much more difficult, because the same selectors are also used by WooCommerce itself, which causes conflicts if you want to sneak additional parameters along. This makes it much easier to make WooCommerce PayPal Payments compatible with other plugins.

**Comment by PoeHaH** [comment]
This issue is related to the one I describe here: https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2322

**Comment by PoeHaH** [comment]
Hi,

Afaik this has not been solved. It feels like the authors deliberately want to keep this functionality closed so tons of (non-Woo) plugins can not properly integrate?

**Comment by isitgeorge** [comment]
This is causing a bit of a headache for some custom functionality I'm working on where the item price is calculated depending on data the customer provides via cart_item_data.

I too am calculating this as part of the `woocommerce_before_calculate_totals` action which is respected by normal card payments via Stripe, but not express checkout options like Apple Pay.

What is the solution/workaround for this? None of the threads I've read actually specify how/when express payments fetch the cart total - is there an earlier action I should be using instead of `woocommerce_before_calculate_totals`?

Appreciate any help!

**Comment by isitgeorge** [comment]
@jonas-hoebenreich Did you find a workaround or solution for this issue in the end, or did you end up disabling Stripe Express Checkouts? Cheers

**Comment by jonas-hoebenreich** [comment]
@isitgeorge yes, I found a workaround that uses a different hook, you can check it out here: https://github.com/wc-donation/wc-donation-platform/blob/main/includes/integrations/express-checkout/class-wcdp-express-checkout.php#L20

**Comment by isitgeorge** [comment]
> [@isitgeorge](https://github.com/isitgeorge) yes, I found a workaround that uses a different hook, you can check it out here: https://github.com/wc-donation/wc-donation-platform/blob/main/includes/integrations/express-checkout/class-wcdp-express-checkout.php#L20

Thanks @jonas-hoebenreich, much appreciated - I will look into it.

I opened another issue here https://github.com/woocommerce/woocommerce-gateway-stripe/issues/4124 which seems to have been triaged as a bug but no updated since March.

**Comment by wjrosa** [comment]
Setting this a duplicate due to the other, more recent, issue linked above.
